programs/quarter_board/communication.c

```diff
@@ -0,0 +1,170 @@
+                    operation_buffer = chCoreAlloc(sizeof(uint8_t) * quantity_of_slaves *
+                                                   MAX_DATA_CHAR_LENGHT);
+                }
+
+                if (data_buffer == NULL) {
+                    data_buffer = chCoreAlloc(sizeof(uint8_t) * quantity_of_slaves * 
+                                              MAX_DATA_CHAR_LENGHT);
+                }
+
+                received_char--;
+                sdPut(&SD1, received_char);
+            } else {
+                my_address = received_char & ~0x80;
+                received_char++;
+                sdPut(&SD1, received_char);
+            }
+        } else if (received_char == 0x0D || received_char == 0x0A) {
+            state = COMMAND_CHAR_0_READ; // communication reset - \r \n
+            sdPut(&SD1, received_char);
+            char_counter = 0;
+            command = 0;
+        } else if ((received_char >= '0' && received_char <= '9') || 
+                   (received_char >= 'A' && received_char <= 'F')) {
+            
+            received_char = char_to_value(received_char);
+
+            if (state <= COMMAND_END) {
+                command |= (received_char << (state * CHAR_BIT_LENGHT));
+
+                sdPut(&SD1, value_to_char(received_char));
+
+                state++;
+
+                if (state == COMMAND_END + 1) {
+                    //commmand_received
+                    for (i = 0; i < (quantity_of_slaves * MAX_DATA_CHAR_LENGHT); i++) {
+                        operation_buffer[i] = NO_OPERATION;
+                        data_buffer[i] = 0;
+                    }
+
+                    data_length = (command >> (OPERATION_BIT_LENGHT + RW_BIT));
+                    rw = (((command >> OPERATION_BIT_LENGHT) & 1) + 1);
+
+                    msgp -> operation = command;
+                    
+                    char_counter = 0;
+
+                    if (rw == SEND_DATA_TO_MASTER) {
+                        msgp = (message_data_t *)chMsgSend(msgThread, (msg_t)msgp);
+                        
+                        for (i = 0; i < data_length; i++) {
+                            data_buffer[i + data_length * my_address] = msgp -> data[i];
+                            operation_buffer[i + data_length * my_address] = SEND_DATA_TO_MASTER;
+                        }
+                    }
+
+                    if (rw == PROCESS_DATA_FROM_MASTER) {
+                        state = PROCESSING_STATE;
+                        
+                        for (i = 0; i < data_length; i++) {
+                            operation_buffer[i + data_length * my_address] = PROCESS_DATA_FROM_MASTER;
+                        }
+                    }
+                }
+            } else {
+                if (state == SENDING_STATE || state == PROCESSING_STATE) {
+                    switch (operation_buffer[char_counter]) {
+                        case NO_OPERATION:              sdPut(&SD1, value_to_char(received_char));

```

Comment:
Please reformat this case so that case is on the line of its own and actual expressions contained in the block are on the next lines, indented as usual. This is actually hard to read and stretches the 80-character limit as well.